### PR TITLE
Create the invalidation endpoint for entity caching

### PR DESCRIPTION
**Pull request by bnjjj**
Create the invalidation endpoint for entity caching

TODO:
- [x] the response to invalidation requests on the endpoint returns the number of invalidated entries. Could we return errors too there? Users would be able to react to errors during invalidation (ex: to retry them)
- [x] the extensions contained a `Vec<InvalidationRequest>` but the endpoint only accept a single `InvalidationRequest`. Do we want to align them?
- [x] the endpoint directly calls the `handle_request` function instead of sending the request to the invalidation task. We may want to have all requests go through that task. That implies passing a channel sender along with the requests, to get notified once the invalidation is finished). It might also requires transmitting the current span too, so the invalidation execution happens under the same trace
- [x] handle the `shared_key` parameter
---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [x] Integration Tests
    - [x] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

[ROUTER-109]: https://apollographql.atlassian.net/browse/ROUTER-109?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apollo-router/src/plugins/cache/invalidation.rs [hunk 1]
```diff
@@ -36,33 +71,53 @@ impl Invalidation {
...
-    ) -> Result<(), BoxError> {
+    ) -> Result<u64, BoxError> {
         if self.enabled {
             let mut sink = self.handle.clone().into_sink();
-            sink.send((origin, requests)).await.map_err(|e| e.message)?;
-        }
+            let (response_tx, mut response_rx) = broadcast::channel(1);
+            sink.send((requests, origin, response_tx))
+                .await
+                .map_err(|e| e.message)?;
```
Comment (garypen) [inline apollo-router/src/plugins/cache/invalidation.rs:96]: Could this be the cause of the BAD_REQUEST that we are seeing in performance testing?

Comment (bnjjj) [reply apollo-router/src/plugins/cache/invalidation.rs:96]: I think so

File: docs/source/configuration/entity-caching.mdx [hunk 2]
```diff
@@ -92,6 +92,11 @@ preview_entity_cache:
     timeout: 5ms # Optional, by default: 2ms
     ttl: 24h # Optional, by default no expiration
 
+  # Configure invalidation endpoint
+  invalidation:
+    listen: 127.0.0.1:4000
+    path: /invalidation
+
```
Comment (garypen) [inline docs/source/configuration/entity-caching.mdx:99]: Do we want to remove this from the PR?

**Review by garypen — CHANGES_REQUESTED** [review]

File: apollo-router/src/configuration/snapshots/apollo_router__configuration__tests__schema_generation.snap [hunk 3]
```diff
@@ -3518,6 +3523,24 @@ expression: "&schema"
...
+          "description": "#/definitions/ListenAddr"
+        },
+        "path": {
+          "description": "Specify on which path you want to listen for invalidation endpoint. (default: /invalidation)",
+          "type": "string"
+        }
+      },
+      "required": [
+        "listen",
+        "path"
```
Comment (IvanGoncharov) [inline apollo-router/src/configuration/snapshots/apollo_router__configuration__tests__schema_generation.snap:3540]: The path description says it has a default value, but it's required at the same time.
Maybe the default should be documented here:
https://github.com/apollographql/router/pull/5614/files#diff-ea34c7fa4e3f6984aae7f7e0d20d22930d5386763c6be9d9fb2f017adf16f746R5601

Comment (bnjjj) [reply apollo-router/src/configuration/snapshots/apollo_router__configuration__tests__schema_generation.snap:3540]: I removed the comment and default implementation

**Review by BrynCooke — APPROVED** [review]

**Review by garypen — APPROVED** [review]

**Pull request merged**